File: src/lexer.cpp
#include "lexer.h"
#include <unordered_map>

namespace fs = std::filesystem;

static const std::unordered_map<char, TokenType> char_tokens{
    {'\0', TokenType::END_OF_FILE}, {'=', TokenType::ASSIGN},
    {'+', TokenType::PLUS},         {'-', TokenType::MINUS},
    {'!', TokenType::BANG},         {'*', TokenType::ASTERISK},
    {'/', TokenType::SLASH},        {'<', TokenType::LESS_THAN},
    {'>', TokenType::GREATER_THAN}, {',', TokenType::COMMA},
    {';', TokenType::SEMICOLON},    {'(', TokenType::LEFT_PAREN},
    {')', TokenType::RIGHT_PAREN},  {'{', TokenType::LEFT_BRACE},
    {'}', TokenType::RIGHT_BRACE},
};

static const std::unordered_map<std::string, TokenType> two_char_tokens{
    {"==", TokenType::EQUALS},
    {"!=", TokenType::NOT_EQUALS},
};

static const std::unordered_map<std::string, TokenType> keywords{
    {"fn", TokenType::FUNCTION},   {"let", TokenType::LET},
    {"true", TokenType::TRUE},     {"false", TokenType::FALSE},
    {"if", TokenType::IF},         {"else", TokenType::ELSE},
    {"return", TokenType::RETURN},
};

static bool isLetter(const char c) {
    return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || c == '_';
}

static bool isNumber(const char c) {
    return std::isdigit(c);
}

static std::string readFile(const fs::path& path) {
    std::ifstream file;
    file.exceptions(std::ifstream::failbit | std::ifstream::badbit);
    file.open(path);

    const auto file_size = static_cast<std::streamsize>(fs::file_size(path));
    std::string file_content(file_size, '\0');
    file.read(file_content.data(), file_size);

    return file_content;
}

Lexer::Lexer(const fs::path& path)
    : input(readFile(path)), ch('\0'), position(0), read_position(0) {
    readChar();
}

char Lexer::peekChar() const {
    if (read_position >= input.length()) {
        return '\0';
    }
    return input[read_position];
}

char Lexer::readChar() {
    ch = peekChar();
    position = read_position;
    read_position++;
    return ch;
}

std::string Lexer::readUntil(const std::function<bool(char)>& accept_char) {
    const auto initial_pos = position;
    while (accept_char(ch)) {
        readChar();
    }
    const auto length = position - initial_pos;
    return input.substr(initial_pos, length);
}

std::optional<Token> Lexer::readCharToken() {
    const auto entry = char_tokens.find(ch);
    if (entry == char_tokens.end()) {
        return std::nullopt;
    }
    const auto type = entry->second;
    std::string value{};
    if (entry->first != '\0') {
        value = std::string{entry->first};
    }
    // Advance one character.
    readChar();
    return Token{type, value};
}

std::optional<Token> Lexer::readTwoCharToken() {
    const char c1 = ch;
    const char c2 = peekChar();
    const std::string value{c1, c2};
    const auto entry = two_char_tokens.find(value);
    if (entry == two_char_tokens.end()) {
        return std::nullopt;
    }
    // Advance two characters.
    readChar();
    readChar();
    return Token{entry->second, value};
}

std::optional<Token> Lexer::readWordToken() {
    const auto word = readUntil(isLetter);
    if (word.empty()) {
        return std::nullopt;
    }
    TokenType type;
    if (const auto entry = keywords.find(word); entry != keywords.end()) {
        type = entry->second;
    } else {
        type = TokenType::IDENTIFIER;
    }
    return Token{type, word};
}

std::optional<Token> Lexer::readNumberToken() {
    const auto num = readUntil(isNumber);
    if (num.empty()) {
        return std::nullopt;
    }
    return Token{TokenType::INT, num};
}

void Lexer::skipWhitespaces() {
    while (std::isspace(ch)) {
        readChar();
    }
}

Token Lexer::nextToken() {
    skipWhitespaces();
    if (const auto token = readTwoCharToken()) {
        return *token;
    }
    if (const auto token = readCharToken()) {
        return *token;
    }
    if (const auto token = readWordToken()) {
        return *token;
    }
    if (const auto token = readNumberToken()) {
        return *token;
    }
    return Token{TokenType::ILLEGAL, ch};
}

File: src/lexer.rs
use crate::token::Token;

#[cfg(test)]
mod tests;

pub struct Lexer<'a> {
    input: &'a str,
    line: usize,
    ch: char,
    current_position: usize,
    read_position: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut lexer = Self {
            input,
            line: 1,
            ch: '\0',
            current_position: 0,
            read_position: 0,
        };
        lexer.read_char();
        lexer
    }

    pub fn line(&self) -> usize {
        self.line
    }

    pub fn next_token(&mut self) -> Token<'a> {
        self.skip_whitespace();
        if self.ch == '\0' {
            return Token::Eof;
        }
        self.lex_fixed_length_token(2)
            .or_else(|| self.lex_fixed_length_token(1))
            .or_else(|| self.lex_word())
            .or_else(|| self.lex_number())
            .expect("Lexer encountered illegal Token")
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            if self.ch == '\n' {
                self.line += 1;
            }
            self.read_char();
        }
    }

    fn lex_fixed_length_token(&mut self, length: usize) -> Option<Token<'a>> {
        let token_symbol = self.peek(length);
        let token = get_token(token_symbol)?;
        for _ in 0..length {
            self.read_char();
        }
        Some(token)
    }

    fn lex_number(&mut self) -> Option<Token<'a>> {
        let number = self.read_while(char::is_ascii_digit)?;
        Some(Token::Integer(number))
    }

    fn lex_word(&mut self) -> Option<Token<'a>> {
        let word = self.read_while(is_letter)?;
        if let Some(keyword) = get_token(&word) {
            return Some(keyword);
        }
        Some(Token::Identifier(word))
    }

    fn read_char(&mut self) {
        let mut chars = self.input.chars();
        if let Some(new_char) = chars.nth(self.read_position) {
            self.ch = new_char;
            self.current_position = self.read_position;
            self.read_position += 1;
            return;
        }
        self.ch = '\0';
        self.current_position = self.read_position;
    }

    pub fn read_while<F>(&mut self, accept_char: F) -> Option<&'a str>
    where
        F: Fn(&char) -> bool,
    {
        let start_pos = self.current_position;
        while accept_char(&self.ch) {
            self.read_char();
        }
        let final_pos = self.current_position;
        if start_pos == final_pos {
            return None;
        }
        Some(&self.input[start_pos..final_pos])
    }

    fn peek(&self, n: usize) -> &'a str {
        let start_pos = self.current_position;
        let final_pos = std::cmp::min(start_pos + n, self.input.len());
        &self.input[start_pos..final_pos]
    }
}

fn is_letter(ch: &char) -> bool {
    ch.is_ascii_alphabetic() || *ch == '_'
}

fn get_token(token_symbol: &str) -> Option<Token<'static>> {
    let token = match token_symbol {
        "=" => Token::Assign,
        "+" => Token::Plus,
        "-" => Token::Minus,
        "!" => Token::Bang,
        "*" => Token::Asterisk,
        "/" => Token::Slash,
        "<" => Token::LessThan,
        ">" => Token::GreaterThan,
        "==" => Token::Equals,
        "!=" => Token::NotEquals,
        "," => Token::Comma,
        ";" => Token::Semicolon,
        "(" => Token::LeftParen,
        ")" => Token::RightParen,
        "{" => Token::LeftBrace,
        "}" => Token::RightBrace,
        "let" => Token::Let,
        "fn" => Token::Function,
        "if" => Token::If,
        "else" => Token::Else,
        "true" => Token::True,
        "false" => Token::False,
        "return" => Token::Return,
        _ => return None,
    };
    Some(token)
}
